src/nvim/change.c

```diff
@@ -0,0 +1,1805 @@
+    buf->b_mod_set = true;
+    buf->b_mod_top = lnum;
+    buf->b_mod_bot = lnume + xtra;
+    buf->b_mod_xlines = xtra;
+  }
+}
+
+/// Changed lines for the current buffer.
+/// Must be called AFTER the change and after mark_adjust().
+/// - mark the buffer changed by calling changed()
+/// - mark the windows on this buffer to be redisplayed
+/// - invalidate cached values
+/// "lnum" is the first line that needs displaying, "lnume" the first line
+/// below the changed lines (BEFORE the change).
+/// When only inserting lines, "lnum" and "lnume" are equal.
+/// Takes care of calling changed() and updating b_mod_*.
+/// Careful: may trigger autocommands that reload the buffer.
+void
+changed_lines(
+    linenr_T lnum,        // first line with change
+    colnr_T col,          // column in first line with change
+    linenr_T lnume,       // line below last changed line
+    long xtra,            // number of extra lines (negative when deleting)
+    bool do_buf_event  // some callers like undo/redo call changed_lines()
+                       // and then increment changedtick *again*. This flag
+                       // allows these callers to send the nvim_buf_lines_event
+                       // events after they're done modifying changedtick.
+)
+{
+  changed_lines_buf(curbuf, lnum, lnume, xtra);
+
+  if (xtra == 0 && curwin->w_p_diff && !diff_internal()) {
+    // When the number of lines doesn't change then mark_adjust() isn't
+    // called and other diff buffers still need to be marked for
+    // displaying.
+    linenr_T wlnum;
+
+    FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
+      if (wp->w_p_diff && wp != curwin) {
+        redraw_win_later(wp, VALID);
+        wlnum = diff_lnum_win(lnum, wp);
+        if (wlnum > 0) {
+          changed_lines_buf(wp->w_buffer, wlnum,
+                            lnume - lnum + wlnum, 0L);
+        }
+      }
+    }
+  }
+
+  changed_common(lnum, col, lnume, xtra);
+
+  if (do_buf_event) {
+    int64_t num_added = (int64_t)(lnume + xtra - lnum);
+    int64_t num_removed = lnume - lnum;
+    buf_updates_send_changes(curbuf, lnum, num_added, num_removed, true);
+  }
+}
+
+/// Called when the changed flag must be reset for buffer "buf".
+/// When "ff" is true also reset 'fileformat'.
+void unchanged(buf_T *buf, int ff)
+{
+  if (buf->b_changed || (ff && file_ff_differs(buf, false))) {
+    buf->b_changed = false;
+    ml_setflags(buf);
+    if (ff) {
+      save_file_ff(buf);
+    }
+    check_status(buf);
+    redraw_tabline = true;
+    need_maketitle = true;  // set window title later
+  }
+  buf_inc_changedtick(buf);
+}
+
+/// Insert string "p" at the cursor position.  Stops at a NUL byte.
+/// Handles Replace mode and multi-byte characters.
+void ins_bytes(char_u *p)
+{
+  ins_bytes_len(p, STRLEN(p));
+}
+
+/// Insert string "p" with length "len" at the cursor position.
+/// Handles Replace mode and multi-byte characters.
+void ins_bytes_len(char_u *p, size_t len)
+{
+  if (has_mbyte) {
```

Comment:
`has_mbyte` is always true in nvim, can remove the branch (3 more cases in this file).